control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+                                 const IR::ControlBlock* controlBlock,
+                                 ReferenceMap* refMap,
+                                 TypeMap* typeMap) {
+  CHECK_NULL(controlBlock);
+  CHECK_NULL(refMap);
+  CHECK_NULL(typeMap);
+
+  auto control = controlBlock->container;
+  CHECK_NULL(control);
+
+  // XXX(seth): These are hardcoded in the BMV2 JSON serializer. Ideally we'd do
+  // something better, but this feature is slated for removal anyway.
+  if (control->name == "ingress" || control->name == "egress") {
+    serializer.addSynthesizedExitTable(control);
+  }
+
+  for (auto declaration : *control->controlLocals) {
+    // Serialize actions and, implicitly, their parameters.
+    if (!declaration->is<IR::P4Action>()) continue;
+    auto action = declaration->to<IR::P4Action>();
+    serializer.addAction(action);
+
+    // Serialize any header field lists which are passed to standard externs in
+    // the action body.
+    // XXX(seth): This is more hardcoded stuff.
+    using CallStatement = IR::MethodCallStatement;
+    forAllMatching<CallStatement>(action, [&](const CallStatement* callStatement) {
+      auto call = callStatement->methodCall;
+      auto instance = P4::MethodInstance::resolve(call, refMap, typeMap);
+      if (!instance->is<P4::ExternFunction>()) return;
+      auto externFunction = instance->to<P4::ExternFunction>();
+      if (externFunction->method->name == "clone") {
+        serializer.addHeaderFieldList("fl", {});
+      } else if (externFunction->method->name == "clone3") {
+        serializeFieldListToPI(serializer, typeMap, "fl", call->arguments->at(2));
+      } else if (externFunction->method->name == "resubmit") {
+        auto name = externalNameFromTypeArgs(call->typeArguments, refMap, "resubmit");
+        serializeFieldListToPI(serializer, typeMap, name, call->arguments->at(0));
+      } else if (externFunction->method->name == "recirculate") {
+        auto name = externalNameFromTypeArgs(call->typeArguments, refMap, "recirculate");
+        serializeFieldListToPI(serializer, typeMap, name, call->arguments->at(0));
+      }
+    });
+  };
+}
+
+static void collectExternSymbols(PISymbolTable& symbols,
+                                 const IR::ExternBlock* externBlock) {
+  CHECK_NULL(externBlock);
+
+  if (externBlock->type->name == "counter") {
+    symbols.add(PISymbolType::COUNTER, externBlock->node->to<IR::IDeclaration>());
+  } else if (externBlock->type->name == "meter") {
+    symbols.add(PISymbolType::METER, externBlock->node->to<IR::IDeclaration>());
+  }
+}
+
+static void serializeExternToPI(P4InfoSerializer& serializer,
+                                const IR::ExternBlock* externBlock,
+                                const TypeMap* typeMap) {
+  CHECK_NULL(externBlock);
+  CHECK_NULL(typeMap);
+
+  if (externBlock->type->name == "counter") {
+    serializer.addCounter(externBlock->node->to<IR::IDeclaration>(),
```

Comment:
I had some difficulty understanding how direct counters/meters are supposed to work, so I removed the partial implementation I had for them from this patch. I'll revisit it once I've had a chance to talk them over with someone who understands them better.